verifier/tests/pkits-4-2.rs

```diff
@@ -0,0 +1,195 @@
+    );
+    let (crls, unix_time) =
+        common::crls_and_time([TRUST_ANCHOR_ROOT_CRL, BAD_NOT_BEFORE_DATE_CA_CRL].as_slice());
+
+    let root = chain.as_ref()[0]
+        .verify_self_signed(unix_time)
+        .expect("Failed verifying root");
+
+    assert_eq!(
+        chain.signing_key(&root, unix_time, crls.as_slice()),
+        Err(Error::CertificateNotYetValid)
+    );
+}
+
+#[test]
+fn invalid_ca_not_before_date_4_2_2() {
+    let (chain, _) = common::chain_and_leaf_key(
+        [
+            TRUST_ANCHOR_ROOT_CERTIFICATE,
+            GOOD_CA_CERT,
+            INVALID_EE_NOT_BEFORE_DATE_TEST_2EE,
+        ]
+        .as_slice(),
+    );
+    let (crls, unix_time) = common::crls_and_time([TRUST_ANCHOR_ROOT_CRL, GOOD_CA_CRL].as_slice());
+
+    let root = chain.as_ref()[0]
+        .verify_self_signed(unix_time)
+        .expect("Failed verifying root");
+
+    assert_eq!(
+        chain.signing_key(&root, unix_time, crls.as_slice()),
+        Err(Error::CertificateNotYetValid)
+    );
+}
+
+#[test]
+fn valid_pre2000_before_date_4_2_4() {
+    let (chain, expected_key) = common::chain_and_leaf_key(
+        [
+            TRUST_ANCHOR_ROOT_CERTIFICATE,
+            GOOD_CA_CERT,
+            VALID_GENERALIZED_TIME_NOT_BEFORE_DATE_TEST_4EE,
+        ]
+        .as_slice(),
+    );
+    let (crls, unix_time) = common::crls_and_time([TRUST_ANCHOR_ROOT_CRL, GOOD_CA_CRL].as_slice());
+
+    let root = chain.as_ref()[0]
+        .verify_self_signed(unix_time)
+        .expect("Failed verifying root");
+
+    let signing_key = chain
+        .signing_key(&root, unix_time, crls.as_slice())
+        .expect("Failed to verify chain");
+
+    assert_eq!(signing_key, expected_key);
+}
+
+#[test]
+fn invalid_ca_not_after_date_4_2_5() {
+    let (chain, _) = common::chain_and_leaf_key(
+        [
+            TRUST_ANCHOR_ROOT_CERTIFICATE,
+            BAD_NOT_AFTER_DATE_CA_CERT,
+            INVALID_CA_NOT_AFTER_DATE_TEST_5EE,
+        ]
+        .as_slice(),
+    );
+    let (crls, unix_time) =
+        common::crls_and_time([TRUST_ANCHOR_ROOT_CRL, BAD_NOT_AFTER_DATE_CA_CRL].as_slice());
+
+    let root = chain.as_ref()[0]
+        .verify_self_signed(unix_time)
+        .expect("Failed verifying root");
+
+    assert_eq!(
+        chain.signing_key(&root, unix_time, crls.as_slice()),
+        Err(Error::CertificateExpired)
+    );
+}
+
+#[test]
+fn invalid_ee_not_after_date_4_2_6() {
+    let (chain, _) = common::chain_and_leaf_key(
+        [
+            TRUST_ANCHOR_ROOT_CERTIFICATE,
+            GOOD_CA_CERT,
+            INVALID_EE_NOT_AFTER_DATE_TEST_6EE,
+        ]
+        .as_slice(),
+    );
+    let (crls, unix_time) = common::crls_and_time([TRUST_ANCHOR_ROOT_CRL, GOOD_CA_CRL].as_slice());
+
+    let root = chain.as_ref()[0]
+        .verify_self_signed(unix_time)
+        .expect("Failed verifying root");
+
+    assert_eq!(
+        chain.signing_key(&root, unix_time, crls.as_slice()),
+        Err(Error::CertificateExpired)
```

Comment:
I'm wonder about the same thing for signature failures as well, though less likely to happen.

It does feel like it would be nice to know *which* certificate was the issue.  Since alloc is available here we could output the subject name as a String in the expired error, same for not before, and possibly others.